pkg/roachprod/vm/gce/create_sdk.go

```diff
@@ -0,0 +1,341 @@
+						DiskType: proto.String("local-ssd"),
+					},
+					Interface: proto.String("NVME"),
+				})
+			}
+		} else {
+			// Persistent disks are network-attached storage.
+			for i := 0; i < providerOpts.PDVolumeCount; i++ {
+				disks = append(disks, &computepb.AttachedDisk{
+					AutoDelete: proto.Bool(true),
+					InitializeParams: &computepb.AttachedDiskInitializeParams{
+						DiskSizeGb: proto.Int64(int64(providerOpts.PDVolumeSize)),
+						DiskType:   proto.String(providerOpts.PDVolumeType),
+					},
+				})
+			}
+		}
+	}
+
+	// Configure networking.
+	region := zone[:len(zone)-2]
+	project := p.GetProject()
+	networkInterfaces := []*computepb.NetworkInterface{
+		{
+			Subnetwork: proto.String(fmt.Sprintf("projects/%s/regions/%s/subnetworks/default", project, region)),
+			AccessConfigs: []*computepb.AccessConfig{
+				{
+					Name: proto.String("External NAT"),
+					Type: proto.String("ONE_TO_ONE_NAT"),
+				},
+			},
+		},
+	}
+
+	// Configure scheduling.
+	scheduling := &computepb.Scheduling{}
+	if providerOpts.preemptible {
+		scheduling.Preemptible = proto.Bool(true)
+		scheduling.OnHostMaintenance = proto.String("TERMINATE")
+		scheduling.AutomaticRestart = proto.Bool(false)
+	} else if providerOpts.UseSpot {
+		scheduling.ProvisioningModel = proto.String("SPOT")
+	} else if providerOpts.TerminateOnMigration {
+		scheduling.OnHostMaintenance = proto.String("TERMINATE")
+	} else {
+		scheduling.OnHostMaintenance = proto.String("MIGRATE")
+	}
+
+	// Configure the service account.
+	var serviceAccounts []*computepb.ServiceAccount
+	sa := providerOpts.ServiceAccount
+	if sa == "" && p.GetProject() == p.defaultProject {
+		sa = providerOpts.defaultServiceAccount
+	}
+	if sa != "" {
+		serviceAccounts = []*computepb.ServiceAccount{
+			{
+				Email:  proto.String(sa),
+				Scopes: []string{"https://www.googleapis.com/auth/cloud-platform"},
+			},
+		}
+	}
+
+	// Configure metadata (including the startup script).
+	metadata := &computepb.Metadata{
+		Items: []*computepb.Items{
+			{
+				Key:   proto.String("startup-script"),
+				Value: proto.String(startupScriptContent),
+			},
+		},
+	}
+
+	// Build the final InstanceProperties.
+	props := &computepb.InstanceProperties{
+		MachineType:       proto.String(providerOpts.MachineType),
+		Disks:             disks,
+		NetworkInterfaces: networkInterfaces,
+		Scheduling:        scheduling,
+		ServiceAccounts:   serviceAccounts,
+		Metadata:          metadata,
+	}
+
+	if providerOpts.MinCPUPlatform != "" {
+		props.MinCpuPlatform = proto.String(providerOpts.MinCPUPlatform)
+	}
+
+	return props, nil
+}
+
+// parseLabelsString converts a "key=value,key2=value2" string into a map.
+func parseLabelsString(labels string) (map[string]string, error) {
+	result := make(map[string]string)
+	if labels == "" {
+		return result, nil
+	}
+
+	pairs := strings.Split(labels, ",")
+	for _, pair := range pairs {
+		kv := strings.SplitN(pair, "=", 2)
```

Comment:
nit: Keys/Values aren't allowed to have `=` right? I think just the normal `Split()` call would catch that case when you check if `len(kv) != 2` while `SplitN` masks it.